src/Microsoft.AspNetCore.Pubternal.Analyzers/PubternalityAnalyzer.cs

```diff
@@ -0,0 +1,240 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+
+namespace Microsoft.AspNetCore.Pubternal.Analyzers
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    public class PubternalityAnalyzer : DiagnosticAnalyzer
+    {
+        public PubternalityAnalyzer()
+        {
+            SupportedDiagnostics = ImmutableArray.Create(new[]
+            {
+                PubturnalityDescriptors.PUB0001,
+                PubturnalityDescriptors.PUB0002
+            });
+        }
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get; }
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.EnableConcurrentExecution();
+
+            context.RegisterCompilationStartAction(analysisContext =>
+            {
+                analysisContext.RegisterSymbolAction(symbolAnalysisContext => AnalyzeTypeUsage(symbolAnalysisContext), SymbolKind.Namespace);
+                analysisContext.RegisterSyntaxNodeAction(syntaxContext => AnalyzeTypeUsage(syntaxContext), SyntaxKind.IdentifierName);
+            });
+        }
+
+        private void AnalyzeTypeUsage(SymbolAnalysisContext context)
+        {
+            var ns = (INamespaceSymbol)context.Symbol;
+            if (ns.Name == "Internal")
```

Comment:
What about `Foo.Bar.Internal.Baz`? This would flag those as non-pubternal right? It should be if any segment of the namespace is the name `Internal`